### PR TITLE
Server Implementation

### PR DESCRIPTION
**Pull request by AriaZhou**

### CODE REVIEW

File: src/Server.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  if (!error) {
+    handler->Start();
+  }
+
+  StartServer();
+}
+
+Server::~Server() { std::cout << "Bye~" << std::endl; }
```
Comment (firewings2515) [inline src/Server.cpp]: add new line at eof

File: src/Server.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ */
+
+#include "Server.h"
+
+#include <boost/asio.hpp>
+#include <boost/bind.hpp>
+#include <boost/enable_shared_from_this.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/shared_ptr.hpp>
+#include <thread>
```
Comment (firewings2515) [inline src/Server.cpp:13]: add new line to separate C++ standard library thread and 3rd party library boost

File: src/Server.cpp (hunk 1 above)
Comment (john0312) [reply src/Server.cpp]: This is reasonable, or ask for clang-format.

File: src/Server.cpp (hunk 2 above)
Comment (john0312) [reply src/Server.cpp:13]: This is a reasonable reviw. You can cite the style guide as well: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: src/Server.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  acceptor.open(endpoint.protocol());
+  acceptor.set_option(boost::asio::ip::tcp::acceptor::reuse_address(true));
+  acceptor.bind(endpoint);
+  acceptor.listen();
+
+  StartServer();
+}
+
+void Server::StartServer() {
+  handler_ptr new_handler(
```
Comment (john0312) [inline src/Server.cpp:42]: Where is this freed?

File: src/Server.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      new HttpRequestHandler(&io_context, &map_handler_list, session_count));
+  ++session_count;
+
+  acceptor.async_accept(*new_handler->GetSocket(),
+                        boost::bind(&Server::HandleAccept, this, new_handler,
+                                    boost::asio::placeholders::error));
+}
+
+void Server::Run() {
+  if (max_thread_count > 1) {
```
Comment (john0312) [inline src/Server.cpp]: You want to support both multithreading and single thread async?
If not, pick one and only use one.

File: src/Server.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    std::vector<std::shared_ptr<std::thread>> threads;
+    for (std::size_t i = 0; i < max_thread_count; i++) {
+      std::shared_ptr<std::thread> thread =
+          std::make_shared<std::thread>([&]() { io_context.run(); });
+      threads.push_back(thread);
+    }
+    for (auto &&thread : threads) {
+      thread->join();
+    }
+  } else {
```
Comment (Geas77) [inline src/Server.cpp]: newline for dividing if-else condition

File: src/Server.cpp (hunk 3 above)
Comment (AriaZhou) [reply src/Server.cpp:42]: It will be freed in HttpRequestHandler.

Comment (john0312) [reply src/Server.cpp:42]: As discussed in class, please plan the ownership of the request handler, and use unique_ptr.

File: src/Server.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  boost::asio::ip::tcp::endpoint endpoint(add.from_string(address), port);
+  acceptor.open(endpoint.protocol());
+  acceptor.set_option(boost::asio::ip::tcp::acceptor::reuse_address(true));
+  acceptor.bind(endpoint);
+  acceptor.listen();
+
+  StartServer();
+}
+
+void Server::StartServer() {
```
Comment (john0312) [inline src/Server.cpp]: [Optional] AcceptOnce() might be a better name.

File: src/Server.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  acceptor.async_accept(*new_handler->GetSocket(),
+                        boost::bind(&Server::HandleAccept, this, new_handler,
+                                    boost::asio::placeholders::error));
+}
+
+void Server::Run() {
+  std::vector<std::shared_ptr<std::thread>> threads;
+  for (std::size_t i = 0; i < max_thread_count; i++) {
+    std::shared_ptr<std::thread> thread =
+        std::make_shared<std::thread>([&]() { io_context.run(); });
```
Comment (john0312) [inline src/Server.cpp]: You can choose to either:
1. [&this]
2. Make io_context a pointer and only pass io_context into the lambda.

File: src/Server.cpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  if (!error) {
+    handler->Start();
+    map_handler_list.insert(std::make_pair(handler->n_count, handler));
+  }
+  AcceptOnce();
+}
+
+void Server::ReleaseHandler(const size_t n_count) {
+  auto it = map_handler_list.find(n_count);
+  map_handler_list.erase(it);
```
Comment (john0312) [inline src/Server.cpp:78]: By the time this returns, destructor of Handler has already been called, see review comment on the Handler's side.

**Review by john0312 — APPROVED** [review]
Approved assuming the comment on destruction of `this` is handled.

**Pull request merged**